base_layer/core/src/mempool/utx_pool/utx_pool_storage.rs

```diff
@@ -0,0 +1,165 @@
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    blocks::block::Block,
+    mempool::utx_pool::{PrioritizedTransaction, UTxPoolConfig, UTxPoolError, UTxPriority},
+    transaction::Transaction,
+    types::Signature,
+};
+use std::{
+    collections::{BTreeMap, HashMap},
+    convert::TryFrom,
+};
+
+/// UTxPool makes use of UTxPoolStorage to provide thread save access to its Hashmap and BTreeMap.
+pub struct UTxPoolStorage {
+    config: UTxPoolConfig,
+    utx_hmap: HashMap<Signature, PrioritizedTransaction>,
+    utx_tree: BTreeMap<UTxPriority, Signature>,
+}
+
+impl UTxPoolStorage {
+    /// Create a new UTxPoolStorage with the specified configuration
+    pub fn new(config: UTxPoolConfig) -> Self {
+        Self {
+            config,
+            utx_hmap: HashMap::new(),
+            utx_tree: BTreeMap::new(),
+        }
+    }
+
+    fn lowest_priority(&self) -> &UTxPriority {
+        self.utx_tree.iter().next().unwrap().0
+    }
+
+    fn remove_lowest_priority_utx(&mut self) {
+        if let Some((priority, sig)) = self.utx_tree.iter().next().map(|(p, s)| (p.clone(), s.clone())) {
+            self.utx_hmap.remove(&sig);
+            self.utx_tree.remove(&priority);
+        }
+    }
+
+    /// Insert a new transaction into the UTxPoolStorage. Low priority transactions will be removed to make space for
+    /// higher priority transactions. The lowest priority transactions will be removed when the maximum capacity is
+    /// reached and the new transaction has a higher priority than the currently stored lowest priority transaction.
+    pub fn insert_utx(&mut self, utx: Transaction) -> Result<(), UTxPoolError> {
+        if utx.body.kernels.len() == 0 {
+            return Err(UTxPoolError::NoKernelsFound);
+        }
+        let utx_key = utx.body.kernels[0].excess_sig.clone();
+        if !self.utx_hmap.contains_key(&utx_key) {
+            let prioritized_utx = PrioritizedTransaction::try_from(utx)?;
+            if self.utx_hmap.len() >= self.config.storage_capacity {
+                if prioritized_utx.priority < *self.lowest_priority() {
+                    return Ok(());
+                }
+                self.remove_lowest_priority_utx();
+            }
+
+            self.utx_tree.insert(prioritized_utx.priority.clone(), utx_key.clone());
+            self.utx_hmap.insert(utx_key, prioritized_utx);
+        }
+        Ok(())
+    }
+
+    /// Insert a set of new transactions into the UTxPoolStorage
+    pub fn insert(&mut self, utxs: Vec<Transaction>) -> Result<(), UTxPoolError> {
```

Comment:
Note that if this method fails halfway through, we don't know which TX caused the problem, and the method exits with the mempool in a "half-half" state. It _might_ be better to drop this and let callers run the iterator themselves.

Reply:
Actually, this is a *mem*pool, so they won't really fail, so I guess this is fine.

Reply:
I would prefer to give the caller the ability to decide how to iterate and add. This method doesn't add any specific functionality but it does prescribe how ownership is moved, which may not be what the caller wants. It restricts their flexibility on when/if to clone for example. I would suggest exposing only the individual `insert` unless there is a very specific need for us to control how it is iterated, or if we need to perform a task before and after iterating

Reply:
This insert function was made private, as I want to use it internally. The insert of a single transaction was written to be idempotent so it will only fail if there is a serious issue.

Reply:
🆒